### PR TITLE
set content-type header in server response

### PR DESCRIPTION
**Pull request by tanmaykm**
The server will now set the content-type header in the response by default, when it is known. I.e.:
- when the server receives a struct response from the API impl that it serializes to JSON, it sets the content type to `application/json`
- when the API impl returns a string, it passes a `text/plain` as response type.

The accept header sent by the client would override the default behavior though. The server applies some simple heuristics to match accepted mime type:
- for JSON responses, a mime of the form `*/json` is chosen if available
- for text responses, a mime of the form `text/*` is chosen if available

If the default heuristics do not suit an implementation, it could always return a `Response` type with the exact content-type header set. If no content type could be matched, the server still passes the response forward in the processing chain, but without a content-type header. This optimistic behavior suits situations where something else in the processing chain could add the header, or the client could be lenient about the content-type.

### CODE REVIEW

File: src/server.jl [hunk 1]
```diff
@@ -80,9 +81,85 @@ function to_param(T, source::Dict, name::String; required::Bool=false, collectio
...
+        return true
+    end
+    return false
+end
+
+# json if it is of the format `.*/json[;.*]` (roughly)
+function is_json_mime_accepted(mime_type::AbstractString)
+    parts = first(strip.(split(mime_type, ";")))
+    parts = strip.(split(parts, "/"))
+    return (length(parts) == 2) && (parts[2] == "json")
```
Comment (pfitzseb) [inline src/server.jl:96]: Should we just regex match on the `split` string instead (same below)?

File: src/server.jl [hunk 2]
```diff
@@ -80,9 +81,85 @@ function to_param(T, source::Dict, name::String; required::Bool=false, collectio
     end
 end
 
+function is_any_mime_accepted(mime_type::AbstractString)
+    parts = first(strip.(split(mime_type, ";")))
```
[the remark this answers is not in the document]
Comment (tanmaykm) [reply src/server.jl:85]: I don't think that's valid. The `;` must be followed by qualifier. E.g. `text/plain; q=0.5, text/html`

File: src/server.jl (hunk 1 above)
Comment (tanmaykm) [reply src/server.jl:96]: Did consider regex.. but `split` seemed simpler/easier to understand to me.

File: src/server.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (tanmaykm) [reply src/server.jl:85]: Here's a link to the [rfc](https://www.rfc-editor.org/rfc/rfc9110.html#section-12.5.1). That does seem to be the format.

**Comment by mortenpi** [comment]
Should we not be taking into account what MIMEs are declared in the spec though? I don't think that information  is currently passed to `server_response` in any way? Should the generator generate another method that returns a tuple of allowed mime types for a given `APIModel` object?

E.g. let's say I have a single `application/json` content object for the response in the spec. If my handler implementation returns e.g. a `String` (-> `text/plain`), shouldn't validation catch that case and issue an internal error instead?

File: src/server.jl [hunk 3]
```diff
@@ -80,9 +81,85 @@ function to_param(T, source::Dict, name::String; required::Bool=false, collectio
...
+    if !hasheader(resp, "X-Request-Id") && hasheader(req, "X-Request-Id")
+        setheader(resp, "X-Request-Id" => header(req, "X-Request-Id"))
+    end
+    return resp
+end
+server_response(req::HTTP.Request, ::Nothing) = server_response(req, "")
+function server_response(req::HTTP.Request, ret; content_type=nothing)
+    !isnothing(content_type) && isempty(content_type) && (content_type = nothing)
+    if isnothing(content_type)
+        accepted_mime_types = strip.(split(header(req, "Accept", "*/*"), ","))
```
[the remark this answers is not in the document]
Comment (tanmaykm) [reply src/server.jl:109]: :+1: made it into a common method in https://github.com/JuliaComputing/OpenAPI.jl/pull/25/commits/eb8076d5f11faa462d1b0c936a0b2b6a317982c9

**Comment by tanmaykm** [comment]
Closing as outdated

**Pull request closed without merging**